pkg/watch/health.go

```diff
@@ -0,0 +1,240 @@
+const POLL_KV_FOR_PODS = 4000
+
+// number of milliseconds between health checks
+const HEALTHCHECK_INTERVAL = 200
+
+// Contains method for watching the consul reality store to
+// track services running on a node. A manager method:
+// WatchHealth tracks the reality store and manages
+// a health checking go routine for each service in the
+// reality store
+
+// PodWatch houses a pod's manifest, a channel to kill the
+// pod's goroutine if the pod is removed from the reality
+// tree, and a bool that indicates whether or not the pod
+// has a running MonitorHealth go routine
+type PodWatch struct {
+	manifest   kp.ManifestResult
+	shutdownCh chan bool
+	hasMonitor bool // indicates whether this pod is being monitored
+	logger     *logging.Logger
+}
+
+// WatchHealth is meant to be a long running go routine.
+// WatchHealth reads from a consul store to determine which
+// services should be running on the host. WatchHealth
+// runs a CheckHealth routine to monitor the health of each
+// service and kills routines for services that should no
+// longer be running.
+func WatchHealth(node, consul, authtoken string, logger *logging.Logger, shutdownCh chan struct{}) {
+	var store kp.Store
+	tochan := make(chan bool)
+	pods := []*PodWatch{}
+	if authtoken != "" {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Token:   authtoken,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	} else {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	}
+
+	go startTimer(tochan, POLL_KV_FOR_PODS)
+	for {
+		select {
+		case _ = <-tochan:
+			// start timer again
+			go startTimer(tochan, POLL_KV_FOR_PODS)
+
+			// check if pods have been added or removed
+			// starts monitor routine for new pods
+			// kills monitor routine for removed pods
+			pods = updateHealthMonitors(store, pods, node, logger)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+// Monitor Health is a go routine that runs as long as the
+// service it is monitoring. Every HEALTHCHECK_INTERVAL it
+// performs a health check and writes that information to
+// consul
+func (p *PodWatch) MonitorHealth(node string, store kp.Store, shutdownCh chan bool) {
+	statusCheck := fmt.Sprintf(kp.GetStatusCheck(), p.manifest.Manifest.StatusPort)
+	tochan := make(chan bool)
+
+	go startTimer(tochan, HEALTHCHECK_INTERVAL)
+	for {
+		select {
+		case _ = <-tochan:
+			p.checkHealth(statusCheck, node, store)
+			go startTimer(tochan, HEALTHCHECK_INTERVAL)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *PodWatch) checkHealth(healthCheck, node string, store kp.Store) {
+	healthstate, res, err := check(healthCheck)
+	health := health.Result{
+		ID:      p.manifest.Manifest.Id,
+		Node:    node,
+		Service: p.manifest.Manifest.Id,
+		Status:  healthstate,
+		Output:  res,
+	}
+
+	err = writeToConsul(health, store)
```

Comment:
Two things here:

First, I don't see any notion of time / TTL on the health information we are sending to Consul and there should be some indication.

Second, and this may indicate we want an explicit expiry on the health state (not just a timestamp indicating when it was generated), but we may want the ability to abstain from sending a redundant health update to Consul if we can avoid the roundtrip. Lets say we are doing health checks every second, and we explicitly set a TTL of 10 seconds on all health reports we make. If on the 2nd check the health has not changed we can avoid rewriting the value to Consul and have a pretty good guarantee that clients won't find the value to be stale.